pytest-lsp: Raise exception if the tested LSP server doesn't start

tombh opened the pull request:
Hey Alex!

First off this is such a great set of tools for the LSP/Pygls ecosystem ❤️ I'm continuously impressed by the comprehensiveness and high quality of your code 🤓 Thanks for making it public. I know this project is young, but very useful nonetheless.

So, in order to be a better, more informed maintainer of Pygls, I've set out to build my own LSP server! It's also going to double as a reasonable starter template for new Pygls projects: https://github.com/tombh/cli-tools-lsp I was looking around for a good way to do end to end tests and of course I ended up here 😏 

I don't know if this PR is at all the right way to go about addressing my little issue. But nevertheless it highlights the blindspot I hit, namely that I didn't get any feedback that my LSP server wasn't even booting. For sure I know this isn't a complete solution to handling a failed LSP server, because it only detects failure within the first 0.1 seconds of startup, but it's useful to catch the common case of the server not even booting.

tombh committed:
pytest-lsp: Error if tested server doesn't start

This isn't a complete solution to handling a failed LSP server, but it's
useful to catch the common case of the server not even booting.

alcarney commented:
>  I'm continuously impressed by the comprehensiveness and high quality of your code

Aww shucks :blush: 

Some general mechanism that make sure that tests don't end up waiting forever on a crashed server will be needed eventually but this looks like it nicely handles your scenario where the server crashes immediately.

Would you mind adding a small note for the changelog in a file at `lib/pytest-lsp/changes/22.fix.rst`? 
I don't see any reason not to push this straight out in a release! :smile:

tombh committed:
pytest-lsp: Return init response from CS.start()

This allows tests to assert against the server's `INITIALIZE` response
without resending the `INITIALIZE` request in the actual test.

tombh committed:
chore: Add changes doc

tombh commented:
Done. I'm curios what the `22` refers to? Will the the minor and patch versions in the version? `0.2.2`?

Also I added another tiny change, I'm not sure what you think? I made `ClientServer.start()` return the server's response, so it can be inserted against in the test body.

Glad to be contributing!

alcarney commented:
> I'm curios what the 22 refers to?

The `22` is the issue number of this PR.

I use [towncrier](https://towncrier.readthedocs.io/en/stable/) to generate changelogs and by following that naming convention it automatically links entries back to the issues/PRs they came from.

I also have a [script](https://github.com/alcarney/lsp-devtools/blob/bb5a755354a8674661746c6ae79b4cb341ab28ca/scripts/make-release.sh#L50) that uses these files to derive what the next version number should be 😄 

> I made ClientServer.start() return the server's response

Looks good!

tombh commented:
Thank you 🎉

Oh nice, I really like the look of towncrier. And your script 🤓 I'm going to look into that for Pygls.

What do you think about [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)? Because that also provides means to filter out relevant changes.

alcarney commented:
I’ve been vaguely aware of it, but haven’t tried them myself. 

Ultimately, I went the towncrier route as it’s probably easier to fixup/amend the change log by updating some files, rather than trying to rewrite commit messages
